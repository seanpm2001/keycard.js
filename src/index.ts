import fetch from 'cross-fetch';
import { sleep } from './utils';

const API_URL = 'https://locahost:3002';
const API_INTERVAL = 30e3;

type KeycardParams = {
  app: string;
  secret?: string;
  URL?: string | URL;
};

type AppKeys = {
  active: string[];
  restricted_monthly: string[];
};

export class Keycard {
  app: string;
  URL: string | URL;
  configured = false;
  private secret: string | undefined;
  private keys: AppKeys = {
    active: [],
    restricted_monthly: []
  };

  constructor(params: KeycardParams) {
    this.app = params.app;
    this.secret = params.secret;
    this.URL = params.URL || API_URL;

    if (!this.secret) {
      console.log('[keycard] No secret provided, skipping keycard.');
      return;
    }
    this.run();
  }

  private async run() {
    try {
      await this.getKeys();
      this.configured = true;
    } catch (e: any) {
      console.log('[keycard]', e.message || e);
    }
    await sleep(API_INTERVAL);
    this.run();
  }

  async getKeys(): Promise<AppKeys> {
    const { app } = this;
    const { result } = await this.callAPI('get_keys');
    if (result?.[app]) {
      // Useful to debug, Uncomment this to see the keys in the console.
      // console.log(
      //   '[keycard] getKeys: Success! Active keys:',
      //   result[app].active.length,
      //   'Restricted monthly keys:',
      //   result[app].restricted_monthly.length
      // );
      this.keys = result[app];
    }
    return this.keys;
  }

  logReq(key: string): { valid: boolean; rateLimited?: boolean } {
    if (key && this.configured) {
      if (!this.keys.active.includes(key)) return { valid: false };
      if (this.keys.restricted_monthly.includes(key)) return { valid: true, rateLimited: true };

      // Increase the total count for this key, but don't wait for it to finish.
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      this.callAPI('log_req', { key }).catch(() => {});
    }
    // If the keycard doesn't receive any keys (incase of a restart), we don't want to block the request.
    return { valid: true, rateLimited: false };
  }

  private async callAPI(method: string, params: any = {}) {
    const { URL, app, secret } = this;
    const result = await fetch(URL, {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
        secret: secret || ''
      },
      body: JSON.stringify({
        jsonrpc: '2.0',
        method,
        params: { app, ...params }
      })
    });
    return result.json();
  }
}
