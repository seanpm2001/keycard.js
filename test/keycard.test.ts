import { describe, expect, it } from 'vitest';
import { Keycard } from '../src';

describe('Test keyCard if no secret is passed', () => {
  let keycard: any = undefined;
  it('should create a new instance of Keycard', () => {
    keycard = new Keycard({
      app: 'snapshot-hub',
      secret: '',
      URL: 'http://localhost:3002'
    });
    expect(keycard).toBeInstanceOf(Keycard);
  });

  it('should return false for configured value', () => {
    expect(keycard.configured).toBe(false);
  });

  it('should return empty arrays for keys', () => {
    expect(keycard.keys).toMatchObject({
      active: [],
      restricted_monthly: []
    });
  });
});
