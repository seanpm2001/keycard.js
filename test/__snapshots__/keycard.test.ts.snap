// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test KeyCard > snapshot 1`] = `
{
  "foo": "bar",
}
`;
